tests/test_ensemble.py

```diff
@@ -0,0 +1,381 @@
+        with self.assertRaises(ValueError):
+            ens = relentless.Ensemble(T=10, P=2.0)
+
+        #test invalid setting of both P and V
+        with self.assertRaises(ValueError):
+            ens = relentless.Ensemble(T=10, P=2.0, V=v_obj)
+
+        #test invalid setting of both mu and N
+        with self.assertRaises(ValueError):
+            ens = relentless.Ensemble(T=10, mu={'A':0.1,'B':0.2}, N={'A':1,'B':2})
+
+        #test setting N as a float
+        with self.assertRaises(TypeError):
+            ens = relentless.Ensemble(T=10, P=1.0, N={'A':1.0})
+
+    def test_set_conjugates(self):
+        """Test setting constant and conjugate/fluctuating parameter values."""
+        v_obj = relentless.Cube(L=3.0)
+        v_obj1 = relentless.Cube(L=4.0)
+
+        #P and mu set
+        ens = relentless.Ensemble(T=10, P=2.0, mu={'A':0.1,'B':0.2})
+        self.assertAlmostEqual(ens.P, 2.0)
+        self.assertEqual(ens.V, None)
+        self.assertDictEqual(ens.mu.todict(), {'A':0.1,'B':0.2})
+        self.assertDictEqual(ens.N.todict(), {'A':None,'B':None})
+        #set conjugate values
+        ens.V = v_obj
+        self.assertAlmostEqual(ens.V.volume, 27.0)
+        ens.N['A'] = 1
+        ens.N['B'] = 2
+        self.assertDictEqual(ens.N.todict(), {'A':1,'B':2})
+        ens.P = 3.0
+        self.assertAlmostEqual(ens.P, 3.0)
+        ens.mu['A'] = 0.3
+        ens.mu['B'] = 0.4
+        self.assertAlmostEqual(ens.mu.todict(), {'A':0.3,'B':0.4})
+
+        #V and N set
+        ens = relentless.Ensemble(T=20, V=v_obj, N={'A':1,'B':2})
+        self.assertEqual(ens.P, None)
+        self.assertAlmostEqual(ens.V.volume, 27.0)
+        self.assertDictEqual(ens.mu.todict(), {'A':None,'B':None})
+        self.assertDictEqual(ens.N.todict(), {'A':1,'B':2})
+        #set constant values
+        ens.P = 1.0
+        self.assertAlmostEqual(ens.P, 1.0)
+        ens.mu['A'] = 0.1
+        ens.mu['B'] = 0.2
+        self.assertDictEqual(ens.mu.todict(), {'A':0.1,'B':0.2})
+        ens.V = v_obj1
+        self.assertAlmostEqual(ens.V.volume, 64.0)
+        ens.N['A'] = 1
+        ens.N['B'] = 2
+        self.assertDictEqual(ens.N.todict(), {'A':1,'B':2})
+
+        #mu and N used for one type each
+        ens = relentless.Ensemble(T=100, V=v_obj, mu={'A':0.1}, N={'B':2})
+        self.assertEqual(ens.P, None)
+        self.assertAlmostEqual(ens.V.volume, 27.0)
+        self.assertDictEqual(ens.mu.todict(), {'A':0.1,'B':None})
+        self.assertDictEqual(ens.N.todict(), {'A':None,'B':2})
+        #set constant values
+        ens.P = 1.0
+        self.assertAlmostEqual(ens.P, 1.0)
+        ens.V = v_obj1
+        self.assertAlmostEqual(ens.V.volume, 64.0)
+        ens.mu['A'] = 0.2
+        ens.mu['B'] = 0.3
+        self.assertDictEqual(ens.mu.todict(), {'A':0.2,'B':0.3})
+        ens.N['A'] = 2
+        ens.N['B'] = 3
+        self.assertDictEqual(ens.N.todict(), {'A':2,'B':3})
+
+    def test_clear(self):
```

Comment:
This test, though, is good to use all these different combos since they will get cleared differently.